src/EllCrv/EllCrv.jl

```diff
@@ -712,38 +703,125 @@ end
 #  Division polynomials
 #
 ################################################################################
-# TODO:
-# - Adjust to Sage/Magma
-# - Implement for long form, see Mazur, Tate, "p-adic sigma function"
-
+"""
+Computes the n-th division polynomial of an elliptic curve defined over a field k following Mazur and Tate. The two-torsion multiplicity flag determines the output. 
+When it is 0 the n-th division polynomial (as a univariate polynomial) without the 2-torsion factor is returned. (i.e. the two-torsion factor is 1).
+When it is 1 the n-th division polynomial (as a multivariate polynomial) is returned (i.e. the two-torsion factor is 2*y + a1*x + a3).
+When it is 2 the n-th division polynomial is returned as a univariate polynomial. (i.e. the two-torsion factor is 4*x^3+b2*x^2+2*b4*x+b6 = (2*y + a1*x + a3)^2).
+"""
+function division_polynomial(E::EllCrv, n::Int, two_torsion_multiplicity::Int, x = PolynomialRing(base_field(E),"x")[2], y = PolynomialRing(parent(x),"y")[2])
```

Comment:
It is rather unidiomatic to make the return type depend on some parameter. It results in rather slow code in every function that calls `division_polynomial`. It would be better if the return type would always be the same, i.e., always univariate or always multivariate. Is there a way to make this possible? I noticed that Magma returns more than one object, but all of them are univariate.

Reply:
I was also wondering about what the best way would be to do this. The thing you usual call a division polynomial is multivariate. But the only time the y comes in, is when n is even because of the 2-torsion factor. (I think the multivariate version is also currently used by the algorithms the students implemented)

But there are situations where you want the other two possibilities: univariate without 2-torsion or univariate with square of two-torsion. These are the two things Magma returns in DivisionPolynomial(E,n).

In Magma the way to get the usual division polynomial is to multiply the one without the two-torsion factor with the multivariate 2-torsion factor that gets returned by TwoTorsionPolynomial(E), but I don't really like that either. (I believe the thing that should be returned is the one you find on wikipedia. A user shouldn't have to figure out it needs to be multiplied by something else).

The way I implemented it right now is the way it was done in Sage. 

Maybe I should split it into two functions? division_polynomial_uv (for univariate) and division_polynomial for the usual one?

I was also wondering if the split for long and short weierstrass form is even that much faster. (The formulas are essentially the same, the short version simply already assumes a_1,..., a3 are 0)

Reply:
Yes, I agree with what you say and I like the idea of having `division_polynomial_univariate` (we nowadays try not to use too many non-standard abbreviations in function names).

Regarding the split, I have no opinion on that one. Your call :)